syntax = "proto3";

option java_package = "com.grpc.demo.service";

message Empty{}

// MESSAGES

message User {
    int32 id = 1;
    string username = 2;
    string first_name = 3;
    string last_name = 4;
    string phone = 5;
    string email = 6;
    int32 role_id = 7;
    bool is_active = 8;
}

message Role{
    int32 id = 1;
    string name = 2;
}

message Donation {
    int32 id = 1;
    string categoria = 2;
    string description = 3;
    int32 cantidad = 4;
    bool eliminado = 5;
}

message Event {
    int32 id = 1;
    string name = 2;
    string description = 3;
    string fecha_hora = 4; // formato ISO 8601
    repeated User users = 5;
}

message LoginRequest {
    string username_or_email = 1;
    string password = 2;
}

message LoginResponse {
    bool success = 1;
    string message = 2;
    User user = 3;
}

message Response {
    bool success = 1;
    string message = 2;
}

// SERVICE

service UserService {
    rpc CreateUser(User) returns(Response);
    rpc UpdateUser(User) returns(Response);
    rpc DeleteUser(User) returns(Response);
    rpc ListUsers(Empty) returns(UserList);
}

message UserList {
    repeated User user = 1;
}

// Login 
service AuthService {
    rpc Login(LoginRequest) returns (LoginResponse);
}

//
service DonationService {
    rpc CreateDonation(Donation) returns(Response);
    rpc UpdateDonation(Donation) returns(Response);
    rpc DeleteDonation(Donation) returns(Response);
    rpc ListDonations(Empty) returns(DonationList);
}

message DonationList {
    repeated Donation donation = 1;
}

//
service EventService {
    rpc CreateEvent(Event) returns(Response);
    rpc UpdateEvent(Event) returns(Response);
    rpc DeleteEvent(Event) returns(Response);
    rpc ListEvents(Empty) returns(EventList);

    //Users
    rpc AddUser(UserEventRequest) returns(Response);
    rpc RemoveUser(UserEventRequest) returns(Response);

    //Donations 
    rpc UseDonations(DonationEventRequest) returns(Response);
}

message EventList {
    repeated Event event = 1;
}

message UserEventRequest {
    int32 event_id = 1;
    int32 user_id = 2;
}

message DonationEventRequest {
    int32 event_id = 1;
    int32 donation_id = 2;
}



